misc/lldb_cruby.py

```diff
@@ -527,10 +527,68 @@ def dump_bits(target, result, page, object_address, end = "\n"):
         check_bits(page, "wb_unprotected_bits", bitmap_index, bitmap_bit, "U"),
         ), end=end, file=result)
 
-def dump_page(debugger, command, result, internal_dict):
+class HeapPageIter:
+    def __init__(self, page, target):
+        self.page = page
+        self.target = target
+        self.start = page.GetChildMemberWithName('start').GetValueAsUnsigned();
+        self.num_slots = page.GetChildMemberWithName('total_slots').unsigned
+        self.counter = 0
+        self.tRBasic = target.FindFirstType("struct RBasic")
+        self.tRValue = target.FindFirstType("struct RVALUE")
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.counter < self.num_slots:
+            obj_addr_i = self.start + (self.counter * self.tRValue.GetByteSize())
+            obj_addr = lldb.SBAddress(obj_addr_i, self.target)
+            slot_info = (self.counter, obj_addr_i, self.target.CreateValueFromAddress("object", obj_addr, self.tRBasic))
+            self.counter += 1
+
+            return slot_info
+        else:
+            raise StopIteration
+
+
+def dump_page_internal(page, target, process, thread, frame, result, debugger, highlight=None):
     if not ('RUBY_Qfalse' in globals()):
         lldb_init(debugger)
 
+    ruby_type_map = ruby_types(debugger)
+
+    freelist = []
+    fl_start = page.GetChildMemberWithName('freelist').GetValueAsUnsigned()
+    tRVALUE = target.FindFirstType("struct RVALUE")
+
+    while fl_start > 0:
+        freelist.append(fl_start)
+        obj_addr = lldb.SBAddress(fl_start, target)
+        obj = target.CreateValueFromAddress("object", obj_addr, tRVALUE)
+        fl_start = obj.GetChildMemberWithName("as").GetChildMemberWithName("free").GetChildMemberWithName("next").GetValueAsUnsigned()
+
+    for (page_index, obj_addr, obj) in HeapPageIter(page, target):
+        dump_bits(target, result, page, obj_addr, end= " ")
+        flags = obj.GetChildMemberWithName('flags').GetValueAsUnsigned()
+        flType = flags & RUBY_T_MASK
+
+        flidx = '   '
+        if flType == RUBY_T_NONE:
+            try:
+                flidx = "%3d" % freelist.index(obj_addr)
+            except ValueError:
+                flidx = '   '
```

Comment:
It looks like objects that are not `T_NONE` will get an empty `{    }` in the output for `flidx`. I think this is some confusing output for objects that are not `T_NONE`.

Reply:
Yup, that was the intended behaviour. It allows you to easily visualise which slots are part of the freelist and which ones aren't. Like this:

```
bits: [     ] T_NONE     [401]: {  2} Addr: 0x1007ebed0 (flags: 0x0)
bits: [     ] T_NONE     [402]: {  1} Addr: 0x1007ebef8 (flags: 0x0)
bits: [     ] T_NONE     [403]: {  0} Addr: 0x1007ebf20 (flags: 0x0)
bits: [LM   ] T_CLASS    [404]: {   } Addr: 0x1007ebf48 (flags: 0x100001062)
bits: [LM   ] T_PAYLOAD  [405]: {   } Addr: 0x1007ebf70 (flags: 0x3077)
```

It's easy to see where slots have been assigned values and where the freelist pointer (and list) is going. Although I suspect we could make it more obvious in the output that `{ }` was the freelist index? How about:

```
bits: [     ] T_NONE     idx: [401] freelist_idx: {  2} Addr: 0x1007ebed0 (flags: 0x0)
bits: [     ] T_NONE     idx: [402] freelist_idx: {  1} Addr: 0x1007ebef8 (flags: 0x0)
bits: [     ] T_NONE     idx: [403] freelist_idx: {  0} Addr: 0x1007ebf20 (flags: 0x0)
bits: [LM   ] T_CLASS    idx: [404] freelist_idx: {   } Addr: 0x1007ebf48 (flags: 0x100001062)
bits: [LM   ] T_PAYLOAD  idx: [405] freelist_idx: {   } Addr: 0x1007ebf70 (flags: 0x3077)
```

the empty `{ }` mirrors the behaviour of `rp` - it shows an empty `[      ]` section when no bits are set

```
bits: [     ]
T_NONE: (RBasic) *$35 = (flags = 0x0000000000000000, klass = 0x00000001007ebea8)
```

Reply:
I like your second one 👍

Reply:
👍 I'll patch that and push it up.